ocaml/xapi/xapi_pool.ml

```diff
@@ -2130,7 +2130,59 @@ let retrieve_wlb_recommendations ~__context = get_opt_recommendations ~__context
 
 let send_test_post = Remote_requests.send_test_post
 
-let certificate_install = Certificates.(pool_install CA_Certificate)
+(* Installs a CA certificate pool-wide as a trust anchor.
+   If the given name is a UUID, use it as the certficate's record uuid,
+   this allows to install self-signed certificates with a given UUID. *)
+let certificate_install ~__context ~name ~cert =
+  let expr = Db_filter.(Eq (Field "uuid", Literal name)) in
+  let refs = Db.Certificate.get_refs_where ~__context ~expr in
+  (* Do not allow users to overwrite certificates installed by xapi
+     To install a self signed:
+       1. write to filesystem
+       2. distribute cert & add to db (here)
+       3. modify the db record so it's a self-signed certificate
+  *)
+  if refs <> [] then
+    raise Api_errors.(Server_error (internal_error, [name])) ;
+
+  (* only names with .pem extension are allowed. *)
+  let uuid, name =
```

Comment:
What if `name` already has a .pem extension? If that's not allowed then we should at least check

Also you didn't add .pem in the None case which is confusing

Reply:
You're completely right, this is just minimal work to make the normal certificate installation to work. There is an ticket that specifically targets proper CA certificate installation (and removal!). In general CA certificates must _not_ be replaced or removed if they are the only trust anchor for a server certificate (host certificate). Self-signed need special management.

Currently the implementation details in the certificates module are being leaked. I'll try to see if I can plug some of the problems, but I don't want to spend too much time since it's going to be replaced soon.

Reply:
For the time being, I've simplified the setup by assuming the name passed to the api is just a name and not the full valid filename. This makes it impossible for a user to rewrite a xapi-installed certificate.